design/deploy-steps.md

```diff
@@ -0,0 +1,150 @@
+and is responsible for the whole deployment process.
+
+The `Image.URL` will no longer be required since some images (such as CoreOS)
+already include the root filesystem.
+
+NOTE: `CustomDeploy` is a single-field structure to allow potential extensions
+in the future. The `UserData` and `NetworkData` will be available to a deploy
+step implementation as part of the Node's `instance_info["configdrive"]`.
+Any required customization is expected to be provided via `UserData` and
+`NetworkData` or be built into the image itself.
+
+The expected workflow is as follows:
+
+1. A user sets `CustomDeploy.Method = "step_name"` and optionally `Image.URL`,
+   `UserData` and `NetworkData`.
+2. BMO sets `Node.DeployInterface = "custom-deploy"` and requests deployment
+   with a custom deploy step `step_name`.
+3. Ironic boots the agent (if not booted for inspection/cleaning already).
+4. Ironic requests the agent to execute the deploy step `step_name` (it is
+   an error to provide a non-existing step), providing it the complete Node
+   object, including `configdrive`.
+5. Ironic waits for the step to finish.
+6. Ironic configures the Node's boot device and clean ups the deployment
+   environment.
+7. Ironic reboots the node into the instance.
+
+On failure (including unsupported deploy step) the node will be moved to the
+`deploy failed` state, and the `LastError` field will be populated.
+
+NOTE: There is currently no way to validate the provided deploy step before
+the deployment starts.
+
+[custom-agent]: https://review.opendev.org/c/openstack/ironic/+/786033
+
+### Risks and Mitigations
+
+The new functionality will be easy to misuse, especially if you're not familiar
+with Ironic. We will need to write the documentation accordingly.
+
+### Work Items
+
+- Add support for accepting a custom deploy step on `BareMetalHost`.
+- Implement [custom-agent][custom-agent] in Ironic.
+
+### Dependencies
+
+- [custom-agent deploy interface in Ironic][custom-agent]
+
+Soft dependencies:
+
+- Fix boot interface validation in Ironic that makes `image_source` always
+  required, even if the deploy interface does not use it
+  ([bug 2008874](https://storyboard.openstack.org/#!/story/2008874)).
+- Delay configdrive rendering in Ironic, so that the deploy step can access
+  the original JSON data, not just the final ISO image
+  ([bug 2008875](https://storyboard.openstack.org/#!/story/2008875)).
+
+### Test Plan
+
+We could potentially add CI jobs with Fedora CoreOS and test that it boots
+successfully. However, this requires deploy ISO support, which is outside of
+the scope of this proposal.
+
+### Upgrade / Downgrade Strategy
+
+N/A
+
+### Version Skew Strategy
+
+N/A
+
+## Drawbacks
+
+- A potential footgun.
+
+## Alternatives
+
+- [Explicit CoreOS support](https://github.com/metal3-io/metal3-docs/pull/177)
+- Expose deploy steps more fully. Would allow greater customization at the
+  expense of a more complex API.
+- Do nothing, keep using live ISO.
```

Comment:
We could add more detail about why using the live-iso approach isn't sufficient, e.g call out the metal3/Ironic features lost by choosing that approach.